### ISSUE TITLE
vso-intellij with TFS over https self-signed certificate

### ISSUE THREAD

**Issue by mmajcica** [Microsoft/vso-intellij#50]
Everyone,

as it took me a bit to understand how to get VSTS plugin for IDEA working with our TFS that is reachable only via https and uses a self signed certificate, I decided to share this experience with others. If interested and facing the same challenge, you can find a comprehensive guide here:

[Working with TFS in IntelliJ IDEA via VSTS Plugin](http://blog.majcica.com/2017/01/02/working-with-tfs-in-intellij-idea-via-vsts-plugin/)

I hope this helps.

Mario

**Comment by leantk** [comment]
@mmajcica,

Thank you so much for this great blog post! We really appreciate you taking the time to document the steps you took to get the plugin working. Not only does this help other users but it also helps us identify pain points that we need to fix. You've been a great help.

Thanks again!
-Leah

**Comment by SilverJan** [comment]
Thank you for the instructions. Please also mention that the cacert file is not always stored in the IntelliJ jre/-directory. For my case, it was the in the jre/-directory of my in PATH system env declared Java installation directory.

See the following MSDN blog post which refers to this issue too: https://blogs.msdn.microsoft.com/tfssetup/2016/12/19/error-ssl-certificate-problem-unable-to-get-local-issuer-certificate/

**Comment by ForNeVeR** [comment]
For now, I'll consider this as an improvement request. While the linked article is very useful (thank you!), the UX of using custom certificates is not so good.

We've improved it partially in #192, there's one last improvement we should add: the TF Everywhere client invoked from the IDE should use the same certificate storage as we use in the IDE itself. It is doable, and after that everything should (almost?) work out-of-the-box (will need to conduct some experiments though).

**Comment by mmajcica** [comment]
@ForNeVeR and how about getting to check SChannel for certs, at least on Windows?
Scm Git does it, can't see why it should not be possible for this plugin.

**Comment by ForNeVeR** [comment]
@mmajcica, sorry, I'm not at all familiar with SChannel. Could you please clarify what exactly could we do about that in our plugin/Azure DevOps/TFVC infrastructure?

**Comment by mmajcica** [comment]
@ForNeVeR basically it's all about curl in this case. You can chose which ssl backend to use and one of those will allow you to interface with schannel, meaning that you will be able to check the ssl store on windows itself. This is handy as most of the self-signed certificates are part of the domain or push via the dc. This means that your windows machine knows about them and you can leverage that instead of expecting the user to import it.
By installing Git Scm you are prompted which ssl backend to use (or later you can set it with git config --global http.sslBackend schannel), so this may be something interesting for this plugin also, at lease in case intellj is running on windows.

**Comment by ForNeVeR** [comment]
Thanks for clarification!

While this is really a good suggestion, I'd argue that it's definitely out-of-scope of the plugin.

It could be addressed by the IntelliJ platform itself (it manages the certificates for all the plugins and other functions usually), but I can't see any way to properly address it with the plugin.

Also I should note that our today's certificate tooling is not _that_ bad comparing to what you've seen when writing your post in 2017. There's a UI that will ask you _on demand_ when Azure DevOps plugin hits any host that has a non-trusted certificate, and you may resolve these on a case-by-case basis (and review your decisions later).

I'd argue that it should solve almost any user issues (aside from a centralized certificate management, which is a valid point, but could be solved by externally managed `cacerts` if **really** necessary).

In other words, currently there're no plans to support any additional certificate storages in the plugin code.

**Comment by mmajcica** [comment]
Thank you. What you are saying is reasonable and good enough. Maybe it may be a suggestion for intellij to start considering schannel on windows. I think this issue can be closed.

**Comment by ForNeVeR** [comment]
We'll close that after we finish our work. As I said, currently there's one last UX-related point remaining: the TF client started from the plugin should use the same certificate storage as a plugin itself.

Thanks for the discussion, and thanks for your post from 2017. I think that it has saved countless brave souls who struggled to make the plugin work with custom certificates over all these years 👍

**Comment by mmajcica** [comment]
brave souls -> most of TFS users :)
